295: Boost the capability of recursive record types

michaelhkay opened the pull request:
Fix issue #295 

The main changes are:

* In place of the special self-reference syntax "..", we now allow recursive use of type aliases, allowing types to be mutually recursive
* We generalise the places that recursive references are allowed, for example the record type used by fn:random-number-generator is now legal
* Subtyping rules for recursive record types are now defined (this was previously a gap in the specification). Acknowledgements to a John Snelson blog post for pointing me in the right direction.

Arithmeticus commented on specifications/xquery-40/src/expressions.xml:
```diff
@@ -6532,6 +6599,52 @@ declare function flatten($tree as tree) as item()* {
+
+                  <note><p>The decision to make recursive types iso-recursive rather than equi-recursive
+                  was made largely because it saves a great deal of implementation complexity without any serious
+                  adverse effects for users. It practice, problems can be avoided by using named item type references
```
"It practice" - > "In practice"

michaelhkay commented:
Following the CG review, I have done a further review of the PR and made some minor changes. These are mostly editorial. However two changes are substantive:

(a) In XSLT (as previously indicated with a TODO) I added `item-type` to the set of component kinds that can be exposed or accepted in a package

(b) In XQuery, I changed the rules for unprefixed names in an item type declaration: these now use the default namespace for elements and types, rather than the null namespace. This makes the rules uniform between the defining occurrence of the name and the references to the name.

I suggest that the CG can now be accepted without further live discussion.

ndw commented:
Per minutes of last week, added the merge without discussion label
